File: src/types/state.ts
import { z } from 'zod';

export const SyncPhaseSchema = z.enum([
  'idle',
  'init',
  'fetching',
  'rewriting',
  'pushing',
  'complete',
  'failed',
]);

export const SyncStateSchema = z.object({
  repo_id: z.string(),
  last_sync_hash: z.string().nullable(),
  last_sync_time: z.date().nullable(),
  sync_phase: SyncPhaseSchema,
  failure_count: z.number().int().min(0),
  last_error: z.string().nullable(),
});

export const CommitMappingSchema = z.object({
  repo_id: z.string(),
  github_hash: z.string(),
  internal_hash: z.string(),
  author_email: z.string().email(),
  rewritten_email: z.string().email(),
  sync_time: z.date(),
});

export const SyncLogStatusSchema = z.enum(['success', 'failed', 'partial']);

export const SyncLogSchema = z.object({
  repo_id: z.string(),
  sync_time: z.date(),
  status: SyncLogStatusSchema,
  commits_synced: z.number().int().min(0),
  commits_rewritten: z.number().int().min(0),
  branches_synced: z.array(z.string()),
  duration_ms: z.number().int().min(0),
  error_message: z.string().nullable(),
  details: z.record(z.unknown()).optional(),
});

export const BackupRecordSchema = z.object({
  repo_id: z.string(),
  backup_tag: z.string(),
  created_at: z.date(),
  expires_at: z.date(),
});

export type SyncPhase = z.infer<typeof SyncPhaseSchema>;
export type SyncState = z.infer<typeof SyncStateSchema>;
export type CommitMapping = z.infer<typeof CommitMappingSchema>;
export type SyncLogStatus = z.infer<typeof SyncLogStatusSchema>;
export type SyncLog = z.infer<typeof SyncLogSchema>;
export type BackupRecord = z.infer<typeof BackupRecordSchema>;

File: src/git_sync/models/state.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class SyncState(BaseModel):
    repo_id: str
    last_sync_hash: str | None = None
    last_sync_time: datetime | None = None
    sync_phase: Literal[
        "idle",
        "init",
        "cloning",
        "fetching",
        "checking",
        "rewriting",
        "pushing",
        "complete",
        "failed",
    ] = "idle"
    failure_count: int = 0
    last_error: str | None = None


class CommitMapping(BaseModel):
    repo_id: str
    github_hash: str
    internal_hash: str
    author_email: str
    rewritten_email: str
    sync_time: datetime


class SyncLog(BaseModel):
    repo_id: str
    sync_time: datetime
    status: Literal["success", "failed", "partial"]
    commits_synced: int
    commits_rewritten: int
    branches_synced: list[str]
    duration_ms: int
    error_message: str | None = None
    details: dict | None = None


class BackupRecord(BaseModel):
    repo_id: str
    backup_tag: str
    created_at: datetime
    expires_at: datetime
